funsor/optimizer.py

```diff
@@ -64,12 +64,12 @@ def binary_to_finitary(op, lhs, rhs):
     return Finitary(op, (lhs, rhs))
 
 
-@associate.register(Finitary, object, tuple)
+@associate.register(Finitary, AssociativeOp, tuple)
```

Comment:
@eb8680 can you review the change here carefully? I believe it is correct, but it is a bit subtle since this pattern will be matched less often after this PR.

Reply:
Seems fine, since the default behavior in `associate` is to `reflect`, preserving the previous behavior in all cases.